backends/magma/ceed-magma.c

```diff
@@ -700,95 +712,378 @@ static int CeedOperatorDestroy_Magma(CeedOperator op) {
-    ierr = CeedBasisApply(op->basis, CEED_NOTRANSPOSE, op->qf->inmode,
-                          &Eu[e*op->Erestrict->elemsize*nc], BEu);
-    CeedChk(ierr);
-    CeedScalar *u_ptr = BEu, *v_ptr = BEv;
-    if (op->qf->inmode & CEED_EVAL_INTERP) { in[0] = u_ptr; u_ptr += Q*nc; }
-    if (op->qf->inmode & CEED_EVAL_GRAD) { in[1] = u_ptr; u_ptr += Q*nc*dim; }
-    if (op->qf->inmode & CEED_EVAL_WEIGHT) { in[4] = u_ptr; u_ptr += Q; }
-    if (op->qf->outmode & CEED_EVAL_INTERP) { out[0] = v_ptr; v_ptr += Q*nc; }
-    if (op->qf->outmode & CEED_EVAL_GRAD) { out[1] = v_ptr; v_ptr += Q*nc*dim; }
-    ierr = CeedQFunctionApply(op->qf, &qd[e*Q*op->qf->qdatasize], Q, in, out);
-    CeedChk(ierr);
-    CeedDebug("\033[31m[CeedOperatorApply_Magma] e=%d: ",e);
-    ierr = CeedBasisApply(op->basis, CEED_TRANSPOSE, op->qf->outmode, BEv,
-                          &Eu[e*op->Erestrict->elemsize*nc]);
-    CeedChk(ierr);
-  }
-  ierr = CeedVectorRestoreArray(etmp, &Eu); CeedChk(ierr);
-  // qdata must be restored
-  ierr = CeedVectorRestoreArray(qdata, (CeedScalar**)&qd); CeedChk(ierr);
-  if (residual) {
-    CeedScalar *res;
-    CeedVectorGetArray(residual, CEED_MEM_HOST, &res);
-    for (int i = 0; i < residual->length; i++)
-      res[i] = (CeedScalar)0;
-    // residual must be restored
-    CeedVectorRestoreArray(residual, &res);
-    ierr = CeedElemRestrictionApply(op->Erestrict, CEED_TRANSPOSE,
-                                    nc, lmode, etmp, residual,
-                                    CEED_REQUEST_IMMEDIATE); CeedChk(ierr);
-  }
-  if (request != CEED_REQUEST_IMMEDIATE && request != CEED_REQUEST_ORDERED)
-    *request = NULL;
+  ierr = CeedCalloc(opmagma->numqin + opmagma->numqout, &opmagma->qdata_alloc);
+  CeedChk(ierr);
+  ierr = CeedCalloc(qf->numinputfields + qf->numoutputfields, &opmagma->qdata);
+  CeedChk(ierr);
+
+  ierr = CeedCalloc(16, &opmagma->indata); CeedChk(ierr);
+  ierr = CeedCalloc(16, &opmagma->outdata); CeedChk(ierr);
+
+  // Set up infield and outfield pointer arrays
+  // Infields
+  ierr = CeedOperatorSetupFields_Magma(qf->inputfields, op->inputfields,
+                                       opmagma->evecs, opmagma->qdata, opmagma->qdata_alloc,
+                                       opmagma->indata, 0, 0, 0,
+                                       qf->numinputfields, Q); CeedChk(ierr);
+
+  // Outfields
+  ierr = CeedOperatorSetupFields_Magma(qf->outputfields, op->outputfields,
+                                       opmagma->evecs, opmagma->qdata, opmagma->qdata_alloc,
+                                       opmagma->indata, qf->numinputfields, opmagma->numein,
+                                       opmagma->numqin, qf->numoutputfields, Q); CeedChk(ierr);
+
+  op->setupdone = 1;
+
   return 0;
 }
 
-static int CeedOperatorGetQData_Magma(CeedOperator op, CeedVector *qdata) {
-  CeedOperator_Magma *impl = op->data;
+static int CeedOperatorApply_Magma(CeedOperator op, CeedVector invec,
```

Comment:
The previous MAGMA backend was a near copy of the Ref backend. I am not sure why OpApply is not working for MAGMA right now.